File: src/services/molecules/AuthFlowManager.ts
/**
 * AuthFlowManager - Molecular Component
 * Combines authentication atoms into a functional authentication flow
 * Groups of atoms bonded together that take on distinct new properties
 */

import { authService, AuthRequest, AuthResponse, OAuthProvider, OAuthToken } from '../atoms/AuthService';

export interface AuthFlowConfig {
  provider: string;
  scopes: string[];
  userId: string;
  agentId?: string;
  forceNew?: boolean;
  isDefault?: boolean;
}

export interface AuthFlowResult {
  success: boolean;
  status: 'completed' | 'pending' | 'connection_required' | 'token_expired' | 'error';
  authUrl?: string;
  token?: string;
  authId?: string;
  error?: string;
}

/**
 * Molecular authentication flow manager
 * Combines atomic auth operations into a cohesive authentication flow
 * Follows single responsibility principle for flow management
 */
export class AuthFlowManager {
  private readonly authService: typeof authService;

  constructor(authServiceInstance = authService) {
    this.authService = authServiceInstance;
  }

  /**
   * Complete authentication flow from start to finish
   * Molecular operation: combines multiple atomic operations
   */
  async executeAuthFlow(config: AuthFlowConfig): Promise<AuthFlowResult> {
    try {
      // Step 1: Check if token already exists
      const tokenExists = await this.authService.checkOAuthTokenExists(
        config.userId,
        config.provider
      );

      if (tokenExists && !config.forceNew) {
        // Token exists, retrieve it
        const tokens = await this.authService.listOAuthTokensForUser(config.userId);
        const existingToken = tokens.find(token => 
          token.providerId === config.provider && 
          (!config.isDefault || token.label === 'default')
        );

        if (existingToken && (!existingToken.expiresAt || new Date(existingToken.expiresAt) > new Date())) {
          return {
            success: true,
            status: 'completed',
            token: existingToken.token,
            authId: existingToken.id,
          };
        }
      }

      // Step 2: Start authentication process
      const authRequest: AuthRequest = {
        provider: config.provider,
        scopes: config.scopes,
        userId: config.userId,
        agentId: config.agentId,
        forceNew: config.forceNew,
        isDefault: config.isDefault,
      };

      const authResponse = await this.authService.authenticate(authRequest);

      // Step 3: Handle response based on status
      switch (authResponse.status) {
        case 'completed':
          return {
            success: true,
            status: 'completed',
            token: authResponse.token,
            authId: authResponse.authId,
          };

        case 'pending':
        case 'connection_required':
          return {
            success: false,
            status: authResponse.status,
            authUrl: authResponse.url,
            authId: authResponse.authId,
          };

        case 'token_expired':
          return {
            success: false,
            status: 'token_expired',
            error: 'Authentication token has expired',
          };

        default:
          return {
            success: false,
            status: 'error',
            error: 'Unknown authentication status',
          };
      }
    } catch (error) {
      return {
        success: false,
        status: 'error',
        error: error instanceof Error ? error.message : 'Authentication flow failed',
      };
    }
  }

  /**
   * Wait for authentication completion
   * Molecular operation: combines polling with status checking
   */
  async waitForAuthCompletion(
    authId: string,
    maxWaitTime: number = 300000, // 5 minutes
    pollInterval: number = 2000 // 2 seconds
  ): Promise<AuthFlowResult> {
    const startTime = Date.now();
    
    while (Date.now() - startTime < maxWaitTime) {
      try {
        // In a real implementation, you would poll the auth status
        // For now, we'll simulate this with a timeout
        await new Promise(resolve => setTimeout(resolve, pollInterval));
        
        // Check if auth is complete (this would be an actual API call)
        // const status = await this.authService.checkAuthStatus(authId);
        
        // For demonstration, we'll assume completion after some time
        if (Date.now() - startTime > 10000) { // 10 seconds for demo
          return {
            success: true,
            status: 'completed',
            token: 'demo-token',
            authId,
          };
        }
      } catch (error) {
        return {
          success: false,
          status: 'error',
          error: error instanceof Error ? error.message : 'Auth completion check failed',
        };
      }
    }

    return {
      success: false,
      status: 'error',
      error: 'Authentication timed out',
    };
  }

  /**
   * Get available OAuth providers
   * Molecular operation: combines provider listing with filtering
   */
  async getAvailableProviders(): Promise<OAuthProvider[]> {
    try {
      const providers = await this.authService.listOAuthProviders();
      return providers.filter(provider => provider.isActive);
    } catch (error) {
      throw new Error(`Failed to get available providers: ${error instanceof Error ? error.message : 'Unknown error'}`);
    }
  }

  /**
   * Revoke authentication for a provider
   * Molecular operation: combines token listing with deletion
   */
  async revokeAuthentication(userId: string, provider: string): Promise<boolean> {
    try {
      const tokens = await this.authService.listOAuthTokensForUser(userId);
      const providerTokens = tokens.filter(token => token.providerId === provider);
      
      let success = true;
      for (const token of providerTokens) {
        const deleted = await this.authService.deleteOAuthToken(token.id, userId);
        if (!deleted) {
          success = false;
        }
      }
      
      return success;
    } catch (error) {
      return false;
    }
  }

  /**
   * Refresh authentication token
   * Molecular operation: combines revocation with re-authentication
   */
  async refreshToken(config: AuthFlowConfig): Promise<AuthFlowResult> {
    // First revoke existing authentication
    await this.revokeAuthentication(config.userId, config.provider);
    
    // Then re-authenticate with forceNew: true
    return this.executeAuthFlow({
      ...config,
      forceNew: true,
    });
  }
}

// Export singleton instance
export const authFlowManager = new AuthFlowManager();

File: src/services/molecules/auth_flow_manager.py
"""
Auth Flow Manager - Molecular Component
Combines authentication atoms into a functional authentication flow
Groups of atoms bonded together that take on distinct new properties
"""

import asyncio
from typing import Dict, List, Optional, Any
from dataclasses import dataclass
from datetime import datetime, timezone
from pydantic import BaseModel, Field

from ..atoms.auth_service import (
    AuthService, 
    AuthRequest, 
    AuthResponse, 
    OAuthProvider, 
    OAuthToken,
    AuthServiceError
)


class AuthFlowConfig(BaseModel):
    """Authentication flow configuration"""
    provider: str
    scopes: List[str]
    user_id: str
    agent_id: Optional[str] = None
    force_new: bool = False
    is_default: bool = False


class AuthFlowResult(BaseModel):
    """Authentication flow result"""
    success: bool
    status: str
    auth_url: Optional[str] = None
    token: Optional[str] = None
    auth_id: Optional[str] = None
    error: Optional[str] = None


class AuthFlowManagerError(Exception):
    """Custom exception for auth flow manager errors"""
    pass


class AuthFlowManager:
    """
    Molecular authentication flow manager
    Combines atomic auth operations into a cohesive authentication flow
    Follows single responsibility principle for flow management
    """
    
    def __init__(self, auth_service: Optional[AuthService] = None):
        self.auth_service = auth_service or AuthService()
    
    async def execute_auth_flow(self, config: AuthFlowConfig) -> AuthFlowResult:
        """
        Complete authentication flow from start to finish
        Molecular operation: combines multiple atomic operations
        """
        try:
            # Step 1: Check if token already exists
            token_exists = await self.auth_service.check_oauth_token_exists(
                config.user_id,
                config.provider
            )
            
            if token_exists and not config.force_new:
                # Token exists, retrieve it
                tokens = await self.auth_service.list_oauth_tokens_for_user(config.user_id)
                existing_token = self._find_existing_token(tokens, config)
                
                if existing_token and self._is_token_valid(existing_token):
                    return AuthFlowResult(
                        success=True,
                        status="completed",
                        token=existing_token.token,
                        auth_id=existing_token.id,
                    )
            
            # Step 2: Start authentication process
            auth_request = AuthRequest(
                provider=config.provider,
                scopes=config.scopes,
                user_id=config.user_id,
                agent_id=config.agent_id,
                force_new=config.force_new,
                is_default=config.is_default,
            )
            
            auth_response = await self.auth_service.authenticate(auth_request)
            
            # Step 3: Handle response based on status
            if auth_response.status == "completed":
                return AuthFlowResult(
                    success=True,
                    status="completed",
                    token=auth_response.token,
                    auth_id=auth_response.auth_id,
                )
            elif auth_response.status in ["pending", "connection_required"]:
                return AuthFlowResult(
                    success=False,
                    status=auth_response.status,
                    auth_url=auth_response.url,
                    auth_id=auth_response.auth_id,
                )
            elif auth_response.status == "token_expired":
                return AuthFlowResult(
                    success=False,
                    status="token_expired",
                    error="Authentication token has expired",
                )
            else:
                return AuthFlowResult(
                    success=False,
                    status="error",
                    error="Unknown authentication status",
                )
                
        except AuthServiceError as e:
            return AuthFlowResult(
                success=False,
                status="error",
                error=str(e),
            )
        except Exception as e:
            return AuthFlowResult(
                success=False,
                status="error",
                error=f"Authentication flow failed: {str(e)}",
            )
    
    async def wait_for_auth_completion(
        self,
        auth_id: str,
        max_wait_time: int = 300000,  # 5 minutes
        poll_interval: int = 2000     # 2 seconds
    ) -> AuthFlowResult:
        """
        Wait for authentication completion
        Molecular operation: combines polling with status checking
        """
        try:
            auth_response = await self.auth_service.wait_for_auth_completion(
                auth_id,
                max_wait_time,
                poll_interval
            )
            
            if auth_response.status == "completed":
                return AuthFlowResult(
                    success=True,
                    status="completed",
                    token=auth_response.token,
                    auth_id=auth_response.auth_id,
                )
            else:
                return AuthFlowResult(
                    success=False,
                    status=auth_response.status,
                    error=f"Authentication not completed: {auth_response.status}",
                )
                
        except AuthServiceError as e:
            return AuthFlowResult(
                success=False,
                status="error",
                error=str(e),
            )
        except Exception as e:
            return AuthFlowResult(
                success=False,
                status="error",
                error=f"Auth completion check failed: {str(e)}",
            )
    
    async def get_available_providers(self) -> List[OAuthProvider]:
        """
        Get available OAuth providers
        Molecular operation: combines provider listing with filtering
        """
        try:
            providers = await self.auth_service.list_oauth_providers()
            return [provider for provider in providers if provider.is_active]
        except Exception as e:
            raise AuthFlowManagerError(f"Failed to get available providers: {str(e)}")
    
    async def revoke_authentication(self, user_id: str, provider: str) -> bool:
        """
        Revoke authentication for a provider
        Molecular operation: combines token listing with deletion
        """
        try:
            tokens = await self.auth_service.list_oauth_tokens_for_user(user_id)
            provider_tokens = [token for token in tokens if token.provider_id == provider]
            
            success = True
            for token in provider_tokens:
                deleted = await self.auth_service.delete_oauth_token(token.id, user_id)
                if not deleted:
                    success = False
            
            return success
        except Exception:
            return False
    
    async def refresh_token(self, config: AuthFlowConfig) -> AuthFlowResult:
        """
        Refresh authentication token
        Molecular operation: combines revocation with re-authentication
        """
        # First revoke existing authentication
        await self.revoke_authentication(config.user_id, config.provider)
        
        # Then re-authenticate with force_new: true
        refresh_config = config.model_copy(update={"force_new": True})
        return await self.execute_auth_flow(refresh_config)
    
    async def get_user_tokens(self, user_id: str, provider: Optional[str] = None) -> List[OAuthToken]:
        """
        Get user tokens with optional provider filtering
        Molecular operation: combines token listing with filtering
        """
        try:
            tokens = await self.auth_service.list_oauth_tokens_for_user(user_id)
            
            if provider:
                tokens = [token for token in tokens if token.provider_id == provider]
            
            return tokens
        except Exception as e:
            raise AuthFlowManagerError(f"Failed to get user tokens: {str(e)}")
    
    async def validate_token(self, user_id: str, provider: str) -> bool:
        """
        Validate if user has valid token for provider
        Molecular operation: combines token checking with validation
        """
        try:
            tokens = await self.get_user_tokens(user_id, provider)
            return any(self._is_token_valid(token) for token in tokens)
        except Exception:
            return False
    
    def _find_existing_token(self, tokens: List[OAuthToken], config: AuthFlowConfig) -> Optional[OAuthToken]:
        """Find existing token based on configuration"""
        for token in tokens:
            if token.provider_id == config.provider:
                if config.is_default and token.label == "default":
                    return token
                elif not config.is_default:
                    return token
        return None
    
    def _is_token_valid(self, token: OAuthToken) -> bool:
        """Check if token is still valid"""
        if not token.expires_at:
            return True  # Token doesn't expire
        
        try:
            expiry_time = datetime.fromisoformat(token.expires_at.replace('Z', '+00:00'))
            return expiry_time > datetime.now(timezone.utc)
        except (ValueError, AttributeError):
            return True  # Assume valid if we can't parse the expiry time


# Singleton instance
auth_flow_manager = AuthFlowManager()
